spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/webservices/client/MetricsClientInterceptor.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2012-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.boot.actuate.metrics.webservices.client;
+
+import java.util.concurrent.TimeUnit;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Timer;
+
+import org.springframework.core.NamedThreadLocal;
+import org.springframework.util.Assert;
+import org.springframework.ws.client.WebServiceClientException;
+import org.springframework.ws.client.support.interceptor.ClientInterceptor;
+import org.springframework.ws.client.support.interceptor.ClientInterceptorAdapter;
+import org.springframework.ws.context.MessageContext;
+
+/**
+ * {@link ClientInterceptor} applied via a {@link MetricsWebServiceTemplateCustomizer} to
+ * record metrics.
+ *
+ * @author Dmytro Nosan
+ * @since 2.1.0
+ */
+class MetricsClientInterceptor extends ClientInterceptorAdapter {
+
+	private static final ThreadLocal<Long> TIMERS = new NamedThreadLocal<>(
+			"WebServices Client Metrics Interceptor");
+
+	private final MeterRegistry meterRegistry;
+
+	private final WebServiceTemplateExchangeTagsProvider tagProvider;
+
+	private final String metricName;
+
+	MetricsClientInterceptor(MeterRegistry meterRegistry,
+			WebServiceTemplateExchangeTagsProvider tagProvider, String metricName) {
+		this.tagProvider = tagProvider;
+		this.meterRegistry = meterRegistry;
+		this.metricName = metricName;
+	}
+
+	@Override
+	public boolean handleRequest(MessageContext messageContext)
+			throws WebServiceClientException {
+		TIMERS.set(System.nanoTime());
```

Comment:
Should use `Timer.start(meterRegistry)` so that you are using the registry's clock for this sample.

Reply:
If `MessageContext` is the same instance in both `handleRequest` and `afterCompletion`, we may be able to employ a `Map` rather than `ThreadLocal`. Not sure if there is any advantage to this.

Reply:
Thank you. Good catch, I will change this.